File: cpp_src/problem2-2-2.cpp
/**
 * @file      problem2-2-2.cpp
 * @author    zhangmin(3273749257@qq.com)
 * @brief     对图像进行高斯滤波
 * @version   0.2
 * @date      2022-04-18
 * @copyright Copyright (c) 2022
 */



#include<iostream>
#include<opencv2/opencv.hpp>
#include<cmath>

using namespace std;
using namespace cv;




//创建一个类进行高斯滤波
/**********************class myGaussianFilter********************************/
/**
* 类名：myGaussianFilter
* 类成员属性：Mat src，要进行高斯滤波的图像
*           int img_h，图像 像素长
*           int img_w，图像 像素宽
*           int gaussianSize_h，高斯核长
*           int gaussianSize_w，高斯核宽
*           double gaussianSigma，高斯核方差
* 类成员函数：
*           myGaussianFilter(Mat& _img,int gaussianSize_h,int gauaasianSize_w,double _sigma)，构造函数，初始化类成员属性
*           void getgaussianArray()，获取高斯核
*           Mat& gaussianMerge()，对图像进行高斯滤波
*/
class myGaussianFilter
{
private:
    /* 图像 */
    Mat src;//待处理的图像
    int img_h;//图像 像素长
    int img_w;//图像 像素长

    /* 高斯核 */
    Mat gaussianKernel;//高斯核
    int gaussianSize_h;//高斯核长
    int gaussianSize_w;//高斯核宽
    double gaussianSigma;//高斯核方差

public:
    //构造函数
    myGaussianFilter(Mat& _img,int h,int w,double _sigma):
        gaussianSigma(_sigma),gaussianSize_h(h),gaussianSize_w(w)
    {
        /* 图像 */
        src=_img;
        //获取图像长和宽
        img_h=src.rows;
        img_w=src.cols;

        /* 初始化高斯核 */
        //(1)获取高斯核的Size
        Size gaussianSize;
        gaussianSize.height=gaussianSize_h;
        gaussianSize.width=gaussianSize_w;
        //(2)创建高斯核
        gaussianKernel.create(gaussianSize,CV_64FC1);
        //(3)初始化高斯核(填充0)
        gaussianKernel=cv::Mat::zeros(gaussianSize,gaussianKernel.type());
    }
    Mat GaussianFilter();

private:
    void getgaussianKernel();
};


/************************获取高斯核*************************/
/**
 * 函数名：  getgaussianArray
 * 参数：   无
 * 返回值： 无
 * 功能：   获取高斯核
 * ***************************************
 * 思路：
 * 1.取gaussianArray[1][1]的位置为中心，坐标（0，0）；
 * 2.坐标为(i-center_i,j-center_j)，用x、y取代，将坐标代入二维高斯函数获取初步的高斯核
 * 3.将高斯核归一化
 */
void myGaussianFilter::getgaussianKernel()
{
    // [1] 准备工作
    // [1-1] 获取中心坐标值
    int center_i= gaussianSize_h / 2;
    int center_j= gaussianSize_w / 2;
    // [1-2] 存储初步得出的高斯核的和
    double gaussianSum=0.0f;
    // [1-3] 存储坐标值
    double x, y;

    // [2] 初步计算高斯核
    // [2-1] 遍历每一个坐标
    for (int i = 0; i < gaussianSize_h; ++i)
    {
        y = pow(i - center_i, 2);
        for (int j = 0; j < gaussianSize_w; ++j)
        {
            // [2-2] 套公式计算每一个坐标的值
            x = pow(j - center_j, 2);
            //因为最后都要归一化的，常数部分可以不计算，也减少了运算量
            double g = exp(-(x + y) / (2 * gaussianSigma*gaussianSigma));
            gaussianKernel.at<double>(i, j) = g;
            gaussianSum += g;//求高斯核的和
        }
    }

    // [3] 归一化
    gaussianKernel = gaussianKernel / gaussianSum;
}



/************************对单通道高斯滤波*************************/
/**
 * 函数名：  GaussianFilter
 * 参数：   无
 * 返回值： Mat dst，高斯模糊后的图像
 * 功能：   对图像进行高斯滤波
 * 补充：   可以是单通道，也可以是三通道
 * ***************************************
 * 思路：
 * 1.主要准备工作
 *  (1)获取高斯核
 *  (2)初始化一个图像，存放高斯滤波后的结果，用到zeros()函数
 *  (3)边界填充src，存储到新图像Newsrc中，用到copyMakeBorder
 *
 * 2.遍历每一个像素点,进行卷积处理
 *  要判断单通道还是三通道
 *  (1)对九宫格进项操作
 *  (1)对求出的和进行范围限定
 *  (2)将求出的和放入图像对应通道中
 *  这里单通道对数值操作，三通道对向量操作
 *
 * 3.返回处理后的图像
 */
Mat myGaussianFilter::GaussianFilter()
{
    // [1] 准备工作
    // [1-1] 获取高斯核
    getgaussianKernel();
    // [1-2] 获取中心坐标值,为高斯滤波做准备
    int center_i = (gaussianKernel.rows - 1) / 2;
    int center_j = (gaussianKernel.cols - 1) / 2;
    // [1-3] 初始化一个图像，存放高斯滤波后的结果
    Mat dst = cv::Mat::zeros(src.size(),src.type());
    // [1-4] 边界填充src，存储到Newsrc中
    Mat Newsrc;
    copyMakeBorder(src, Newsrc, center_i, center_i, center_j, center_j, cv::BORDER_REPLICATE);//边界复制

    // [2] 高斯滤波
    // [2-1] 遍历每一个像素点
    for (int i = center_i; i < src.rows + center_i;++i)
    {
        for (int j = center_j; j < src.cols + center_j; ++j)
        {
           // [2-2] 卷积处理
           double sum[3] = { 0 };//存放卷积核与图像九宫格相乘的和(单通道或三通道)
           // 嵌套两层for循环遍历高斯掩膜
           for (int r = -center_i; r <= center_i; ++r)
           {
               for (int c = -center_j; c <= center_j; ++c)
               {
                   /* 通道数为1 */
                   if (src.channels() == 1)
                       sum[0] = sum[0] + Newsrc.at<uchar>(i + r, j + c) * gaussianKernel.at<double>(r + center_i, c + center_j);

                   /* 通道数为3 */
                   else if (src.channels() == 3)
                   {
                       // 利用Vec3b向量对三通道进行处理
                       Vec3b rgb = Newsrc.at<cv::Vec3b>(i+r,j + c);
                       sum[0] = sum[0] + rgb[0] * gaussianKernel.at<double>(r + center_i, c + center_j);//B
                       sum[1] = sum[1] + rgb[1] * gaussianKernel.at<double>(r + center_i, c + center_j);//G
                       sum[2] = sum[2] + rgb[2] * gaussianKernel.at<double>(r + center_i, c + center_j);//R
                   }
               }
           }

           // [2-3] 对值进行限制0-255
           for (int k = 0; k < src.channels(); ++k)
           {
               if (sum[k] < 0)
                   sum[k] = 0;
               else if (sum[k]>255)
                   sum[k] = 255;
           }


           // [2-4] 将处理完后的值放入对应通道
           /* 通道数为1 */
           if (src.channels() == 1)
               dst.at<uchar>(i - center_i, j - center_j) = static_cast<uchar>(sum[0]);

           /* 通道数为3 */
           else if (src.channels() == 3)
           {
               Vec3b rgb = { static_cast<uchar>(sum[0]), static_cast<uchar>(sum[1]), static_cast<uchar>(sum[2]) };//这里进行了类型转换
               dst.at<Vec3b>(i-center_i, j-center_j) = rgb;
           }
        }
    }

    // [3] 返回高斯模糊后的图像
    return dst;
}




int main()
{
    // 1.读取图像
    Mat img=imread("D:/Picture/lena.jpg");
    if(img.empty())//未获取到图像
    {
        cout << "can't read this image!" << endl;
        return 0;
    }

    // 2.实现高斯滤波
    myGaussianFilter test(img,3,3,sqrt(2));
    Mat temp=test.GaussianFilter();

    // 3.显示图像
    imshow("src",img);
    imshow("GaussianFilter",temp);

    // 4.等待按键操作
    waitKey(0);
    return 0;
}

File: py_src/problem2-2-2.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2022/4/18 20:48
# @Author  : zhangmin
# @File    : problem2-2-2.py
# @Brief   :
# @Version : 0.1

import cv2
import numpy as np

'''
函数名：    gaussian_filter
参数：     image          原图像
          K_size         高斯核尺寸，size*size
          sigma          高斯核方差
返回值：    gaussian_img   高斯滤波后的图像
功能：     对图像进行高斯滤波
**************************************
思路：
1.准备工作：获取图像长，宽，通道数
2.图像填充，补零操作
3.高斯核计算
4.进行高斯滤波
  4-1.注意对矩阵元素值的限定
  4-2.将矩阵元素类型转换为uint8
5.返回高斯滤波后的图像
'''


def gaussian_filter(img, K_size=3, sigma=2):
    # [1]获取图像长，宽，通道数
    H, W, C = img.shape

    # [2]图像填充，补零操作
    # [2-1]计算需要添加的层数，同时也方便后面的卷积操作
    center = K_size // 2
    # [2-2]创建一个三维矩阵，存放高斯滤波后的图像
    gaussian_img = np.zeros((H + center * 2, W + center * 2, C), dtype=np.float)
    # [2-3]将原图像copy到该矩阵中
    gaussian_img[center: center + H, center: center + W] = img.copy().astype(np.float)

    # [3]高斯核计算
    # [3-1]初始化高斯核，大小为size*size的二维矩阵
    gaussian_kernel = np.zeros((K_size, K_size), dtype=np.float)
    # [3-2]将坐标代入，初步计算出高斯核
    for x in range(-center, center):
        for y in range(-center, center):
            gaussian_kernel[y + center, x + center] = np.exp(-(x ** 2 + y ** 2) / (2 * (sigma ** 2)))
    gaussian_kernel /= (2 * np.pi * sigma * sigma)
    # [3-3]归一化
    gaussian_kernel /= gaussian_kernel.sum()

    # [4]进行高斯滤波
    # [4-1]将gaussian_img复制一份，方便卷积操作
    tmp = gaussian_img.copy()
    # [4-2]遍历所有像素点(三通道)，卷积操作
    for x in range(H):
        for y in range(W):
            for c in range(C):
                gaussian_img[center + x, center + y, c] = np.sum(gaussian_kernel * tmp[x: x + K_size, y: y + K_size, c])
    # [4-3]对矩阵的元素值进行范围限定，0~255
    gaussian_img = np.clip(gaussian_img, 0, 255)
    # [4-4]将矩阵元素类型转换为uint8
    gaussian_img = gaussian_img[center: center + H, center: center + W].astype(np.uint8)

    # [5]返回高斯滤波后的图像
    return gaussian_img


if __name__ == '__main__':
    # 1.读取图像
    img = cv2.imread("D:/Picture/lena.jpg")

    # 2.高斯滤波
    gaussian_img = gaussian_filter(img, K_size=3, sigma=2)

    # 3.展示图像
    cv2.imshow("RGB", img)
    cv2.imshow("Gaussian", gaussian_img)

    # 4.等待按键操作
    cv2.waitKey(0)
